src/SqlUtils.ts

```diff
@@ -5,52 +5,96 @@ import Constants from "./Constants";
 import SqlConnectionConfig from "./SqlConnectionConfig";
 
 export default class SqlUtils {
+
+    /**
+     * Tries connection to server to determine if client IP address is restricted by the firewall.
+     * First tries with master connection, and then with user DB if first one fails.
+     * Returns the client IP address if firewall restriction is present, or an empty string if connection succeeds. Throws otherwise.
+     */
     static async detectIPAddress(connectionConfig: SqlConnectionConfig): Promise<string> {
-        let ipAddress = '';
+        // First try connection to master
+        let ipAddress = await this.tryConnection(connectionConfig, true, true);
+        if (ipAddress !== undefined) {
+            return ipAddress;
+        }
 
+        // Retry connection with user DB
+        ipAddress = await this.tryConnection(connectionConfig, false, false);
+        if (ipAddress !== undefined) {
+            return ipAddress;
+        }
+        else {
+            throw new Error(`Failed to detect IP address.`);
+        }
+    }
+
+    /**
+     * Tries connection with the specified configuration.
+     * @param config Configuration for the connection.
+     * @param useMaster If true, uses "master" instead of the database specified in @param config. Every other config remains the same.
+     * @param suppressError If true, will not throw connection errors. Non-connection errors will always be thrown.
+     * @returns If connection succeeds, returns empty string. If connection fails due to firewall rule, returns the client's IP address.
+     * Otherwise returns undefined if there were errors but were suppressed by @param suppressError.
```

Comment:
it's a little confusing that the return type has 3 different states that the caller needs to know: ipAddress = connection failed due to firewall rule, empty string = successfully connected, and undefined = error when connecting.

Can this return an object that has properties that make it more clear what happened? Maybe something like
```
{
    success: boolean,
    ipAddress?: string
}
```

Reply:
Added a ConnectionResult interface